qiskit/circuit/quantumcircuit.py

```diff
@@ -1210,6 +1174,8 @@ def append(
             )
         if not isinstance(instruction, Instruction) and hasattr(instruction, "to_instruction"):
             instruction = instruction.to_instruction()
+        if not isinstance(instruction, Instruction):
+            raise CircuitError("object is not an Instruction.")
```

Comment:
I'd be tempted to nest it this way to avoid the double `isinstance` call but I'm not sure it'll make a big difference here

```suggestion
        if not isinstance(instruction, Instruction):
            if hasattr(instruction, "to_instruction"):
                instruction = instruction.to_instruction()
            else:
                raise CircuitError("object is not an Instruction.")
```

Reply:
This is certainly true in an ideal world - I left it this way because it matched the current implementation, which doesn't trust that `to_instruction()` will return an `Instruction` instance (e.g. if there's a protocol clash where we're accidentally calling some random coincidental method on somebody's clash).

I can change it if you prefer - it depends which situation we care more about.  (For reference, `isinstance` is a ~50ns call.)